examples/sha256.rs

```diff
@@ -68,14 +65,22 @@ impl<F: LurkField> CoCircuit<F> for Sha256Coprocessor<F> {
             })
             .collect();
 
-        let result_ptr = enumerate(nums).try_fold(g.nil_ptr.clone(), |acc, (i, num)| {
-            let ptr = AllocatedPtr::alloc_tag(
-                &mut cs.namespace(|| format!("limb_value_{i}")),
-                ExprTag::Num.to_field(),
-                num,
-            )?;
-            AllocatedPtr::construct_cons(cs.namespace(|| format!("limb_{i}")), g, store, &ptr, &acc)
-        })?;
+        let expected_nums: Vec<AllocatedNum<_>> = (0..2)
+            .map(|i| {
+                AllocatedNum::alloc(cs.namespace(|| format!("allocating {i}")), || {
+                    Ok(F::from_u128(self.expected[i]))
+                })
+                .unwrap()
+            })
+            .collect();
+        
+        cs.enforce(
+            || "enforce num 1",
+            |lc| lc + CS::one(),
+            |_| BNum::from(nums[0].clone()).lc(F::from(1)),
+            |_| BNum::from(expected_nums[0].clone()).lc(F::from(1)),
+        );
```

Comment:
This appears to allocate two numbers and then constrain only one of them. If that's what is happening, then the second number would be unconstrained, which is a soundness bug.

Reply:
Agh you're right, I added the first constraint and forgot to add the second once I got it all to work.

Reply:
Understandable. Hopefully this sensitizes you to the general issue and also to the best way to think in order to avoid it by default.

Reply:
It's generally much preferable to minimize the review/audit/possibility-of-error surface area by moving constraint logic into library components and reusing them. The more you splatter ad-hoc allocations and constraints into leaf code, the more likely you are to introduce potentially subtle problems. The best case is outcome is that it consumes much more time in ongoing review and maintenance to prevent that.

Reply:
> It would be less error prone to use circuit function that both allocates and constrains, like for example `circuit::gadgets::data::allocate_constant()`. If you used that on both numbers, then I think this apparent problem would be avoided.

Ahh this is good to know, I'll replace the `enforce`s with that.